### PR TITLE
Problem: add inner problem names, bounds and hierarchical flag

### PR DESCRIPTION
**Pull request by Doresic**
Added a hierarchical flag to `Problem` indicating it is a hierarchical problem. In that case, it contains `inner_x_names`, `inner_lb `, and `inner_ub` of interpretable inner parameters (scaling, offset, noise). I exclude non-interpretable ones (spline & ordinal) as there can easily be too many of them and they are truly hard to interpret by themselves.

### CODE REVIEW

**Review by dweindl — COMMENTED** [review]
Hi, how about

```python

class HierarchicalProblem(Problem):
    ...
    def __init__(...):
        inner_{x_names,lb,ub} = ...
```
?

**Review by FFroehlich — APPROVED** [review]

**Comment by Doresic** [comment]
> Hi, how about
> 
> ```python
> class HierarchicalProblem(Problem):
>     ...
>     def __init__(...):
>         inner_{x_names,lb,ub} = ...
> ```
> 
> ?

Hmm yeah, that'll be nicer. Will change it up.

File: pypesto/hierarchical/base_problem.py [hunk 1]
```diff
@@ -74,7 +74,7 @@ def get_interpretable_x_ids(self) -> list[str]:
 
         Interpretable parameters need to be easily interpretable by the user.
         Examples are scaling factors, offsets, or noise parameters. An example
-        for a non-interpretable inner parameters are spline heights of spline
+        of a non-interpretable inner parameters are spline heights of spline
```
Comment (dweindl) [inline pypesto/hierarchical/base_problem.py:77]: is/are

File: pypesto/hierarchical/inner_calculator_collector.py [hunk 2]
```diff
@@ -248,7 +248,14 @@ def get_inner_par_ids(self) -> list[str]:
         ]
 
     def get_interpretable_inner_par_ids(self) -> list[str]:
-        """Return the ids of interpretable inner parameters of all inner problems."""
+        """Return the ids of interpretable inner parameters of all inner problems.
+
+        Interpretable parameters need to be easily interpretable by the user.
+        Examples are scaling factors, offsets, or noise parameters. An example
+        of a non-interpretable inner parameters are spline heights of spline
```
Comment (dweindl) [inline pypesto/hierarchical/inner_calculator_collector.py:255]: is/are

File: pypesto/problem/base.py [hunk 3]
```diff
@@ -72,6 +72,10 @@ class Problem:
     startpoint_method:
         Method for how to choose start points. ``False`` means the optimizer
         does not require start points, e.g. for the ``PyswarmOptimizer``.
+    hierarchical:
+        A flag indicating whether the problem is hierarchical. If True,
+        the objective function's calculator will collect the objective
+        values of all inner optimization problems.
```
Comment (dweindl) [inline pypesto/problem/base.py:78]: would remove, just use `isinstance` elsewhere

File: pypesto/problem/base.py [hunk 4]
```diff
@@ -166,6 +171,9 @@ def __init__(
         # convert startpoint method to class instance
         self.startpoint_method = to_startpoint_method(startpoint_method)
 
+        # a flag indicating whether the problem is hierarchical
+        self.hierarchical = hierarchical
+
```
Comment (dweindl) [inline pypesto/problem/base.py:176]:
```suggestion
```



would remove, just use isinstance elsewhere

File: pypesto/problem/hierarchical.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    """
+
+    def __init__(
+        self,
+        inner_x_names: Optional[Iterable[str]] = None,
+        inner_lb: Optional[Union[np.ndarray, List[float]]] = None,
+        inner_ub: Optional[Union[np.ndarray, List[float]]] = None,
+        **problem_kwargs: dict,
+    ):
+        super().__init__(**problem_kwargs, hierarchical=True)
```
Comment (dweindl) [inline pypesto/problem/hierarchical.py:46]:
```suggestion
        super().__init__(**problem_kwargs)
```

File: pypesto/problem/hierarchical.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+        if inner_x_names is None:
+            inner_x_names = (
+                self.objective.calculator.get_interpretable_inner_par_ids()
+            )
+        if len(set(inner_x_names)) != len(inner_x_names):
+            raise ValueError("Parameter names inner_x_names must be unique")
+        self.inner_x_names = inner_x_names
+
+        if inner_lb is None or inner_ub is None:
```
Comment (dweindl) [inline pypesto/problem/hierarchical.py:56]: Check that bound dimensions match number of parameters?

File: pypesto/problem/base.py [hunk 7]
```diff
@@ -103,6 +103,7 @@ def __init__(
         ub_init: Union[np.ndarray, List[float], None] = None,
         copy_objective: bool = True,
         startpoint_method: Union[StartpointMethod, Callable, bool] = None,
+        hierarchical: bool = False,
```
Comment (dweindl) [inline pypesto/problem/base.py:106]: still needed?

**Review by dweindl — APPROVED** [review]
:+1:

**Review by stephanmg — APPROVED** [review]

File: pypesto/hierarchical/inner_calculator_collector.py [hunk 8]
```diff
@@ -248,7 +248,14 @@ def get_inner_par_ids(self) -> list[str]:
...
     def get_interpretable_inner_par_ids(self) -> list[str]:
-        """Return the ids of interpretable inner parameters of all inner problems."""
+        """Return the ids of interpretable inner parameters of all inner problems.
+
+        Interpretable parameters need to be easily interpretable by the user.
+        Examples are scaling factors, offsets, or noise parameters. An example
+        of non-interpretable inner parameters is the spline heights of spline
+        approximation for semiquantitative data. It is challenging to interpret
+        the meaning of these parameters based solely on their value.
+        """
```
Comment (PaulJonasJost) [inline pypesto/hierarchical/inner_calculator_collector.py:258]: Yo could link to the documentation of the base problem, but not necessary. Would save some lines and guarantee a common description.

**Review by PaulJonasJost — APPROVED** [review]
Looks good to me, thanks :)

**Pull request merged**